### ISSUE TITLE
Last jedi release (0.18.0) is incompatible with ipython (7.19 and 7.18 tested); reason - column arg was deprecated, and now removed

### ISSUE THREAD

**Issue by DaniilAnichin** [ipython/ipython#12740]
Relevant traceback reads as follows: 
```
  File "../venv/lib/python3.8/site-packages/IPython/core/completer.py", line 2029, in _complete
    completions = self._jedi_matches(
  File "../venv/lib/python3.8/site-packages/IPython/core/completer.py", line 1373, in _jedi_matches
    interpreter = jedi.Interpreter(
  File "../venv/lib/python3.8/site-packages/jedi/api/__init__.py", line 725, in __init__
    super().__init__(code, environment=environment,
TypeError: __init__() got an unexpected keyword argument 'column'
```

sys info: 
```
{'commit_hash': '62779a198',
 'commit_source': 'installation',
 'default_encoding': 'utf-8',
 'ipython_path': '../venv/lib/python3.8/site-packages/IPython',
 'ipython_version': '7.18.0',
 'os_name': 'posix',
 'platform': 'Linux-4.15.0-128-generic-x86_64-with-glibc2.17',
 'sys_executable': '../venv/bin/python',
 'sys_platform': 'linux',
 'sys_version': '3.8.5 (default, Jul 20 2020, 19:50:14) \n[GCC 5.4.0 20160609]'}
```
same reported in jedi repo too

**Comment by benwu95** [comment]
https://github.com/ipython/ipython/commit/dcd9dc90aee7e4c5c52ce44c18e7518934790612

The code has been already updated, but `7.19.0` did not include this. :(

**Comment by davidhalter** [comment]
As a temporary fix for anyone just trying to get things working again:

```
pip install jedi==0.17.2
```

----

It would be really nice if you could quickly release a 7.19.1.

Sorry for that. I did not realize that IPython with that fix was not released yet. I usually test IPython completions before doing a Jedi release, but not this time :/. It will probably also not happen in the future anymore, because I'm going to release Jedi 1.0 soon, so this is probably the last time for a long time that you have to deal with deprecations in Jedi.

Still wish you a happy Christmas!

**Comment by davidhalter** [comment]
By the way, a `7.19.1` release with the dependency `jedi<0.18.0` would also suffice.

**Comment by jkryanchou** [comment]
@davidhalter Thanks for your solution. it finally work.

**Comment by SamuelDSR** [comment]
I had the same problem with ipython and thanks to the solution of @davidhalter , it works again. :+1:

**Comment by digglife** [comment]
Thank you for opening this issue.  Happy holiday!

**Comment by yurzo** [comment]
Posted a pr to pin the dependency as suggested: #12746

**Comment by bremme** [comment]
Thanks a lot. I thought my shell was broken. Every time when I tried to use tab completion in IPython it crashed. Glad I found a solution.

**Comment by aprilahijriyan** [comment]
Finally found a solution! thank you @davidhalter 🎉

**Comment by tik9** [comment]
When would the problem be fixed so that every Jedi version is compatible with ipython?

**Comment by NeilGirdhar** [comment]
@tik9 Does installing from master work for you?

> pip install git+https://github.com/ipython/ipython

If so, whenever ipython releases a new version, it will be fixed for everyone.

**Comment by tik9** [comment]
@NeilGirdhar , when doing
`pip` install git+https://github.com/ipython/ipython`

I get:
`Collecting git+https://github.com/ipython/ipython
 Cloning https://github.com/ipython/ipython to /tmp/pip-req-build-ci5vn99q`
I am on the wrong track, can you help?

**Comment by omarish** [comment]
Got this issue as well. Pinning `jedi==0.17.2` worked for me, thanks @davidhalter.

**Comment by stefanschlipfi** [comment]
> As a temporary fix for anyone just trying to get things working again:
> 
> ```
> pip install jedi==0.17.2
> ```
> 
> It would be really nice if you could quickly release a 7.19.1. (It's already fixed on master).
> 
> Sorry for that. I did not realize that IPython with that fix was not released yet. I usually test IPython completions before doing a Jedi release, but not this time :/. It will probably also not happen in the future anymore, because I'm going to release Jedi 1.0 soon, so this is probably the last time for a long time that you have to deal with deprecations in Jedi.
> 
> Still wish you a Merry Christmas!

Thank you for your solution.
I got the same issue and IPython worked after I installed Jedi

**Comment by nicksama88** [comment]
installing 0.17.2 of jedi also worked for me.  Was chasing my tail trying to figure out why it wasn't working in a new virtual environment, glad to have found this!  Hope the fix is out soon.  Luckily this showed up as the top link in Google for me when searching "ipython init got an unexpected keyword argument 'column'".

**Comment by bl-ue** [comment]
This works:
```console
$ pip install 'jedi<0.18'
```

**Comment by ABODFTW** [comment]
Thanks @bl-ue

For reference, the autocomplete on my Jupyter notebook wasn't working and I was getting this same error on the terminal.

And now it's working as expected after downgrading jedi to 0.17.2 by just executing the command @bl-ue mentioned.

**Comment by BallsyWalnuts** [comment]
Confirmed that downgrading jedi to `0.17.2` fixed the issue for me as well. Thanks @davidhalter

**Comment by gabrielebndn** [comment]
Hi everyone, I've stumbled on this issue while creating docker images for myself.
If I understand correctly from [this comment](https://github.com/ipython/ipython/issues/12740#issuecomment-751273086) the problem has already been solved on master, but so far no release including the fix has been issued.
In order to better organize my own work, may I know when do you plan a new release? Somebody [here](https://github.com/ipython/ipython/issues/12740#issuecomment-751273584) was suggesting to quickly release 7.19.1, is it still an option?

**Comment by HosseinDahaee** [comment]
> As a temporary fix for anyone just trying to get things working again:
> 
> ```
> pip install jedi==0.17.2
> ```
> 
> It would be really nice if you could quickly release a 7.19.1. (It's already fixed on master).
> 
> Sorry for that. I did not realize that IPython with that fix was not released yet. I usually test IPython completions before doing a Jedi release, but not this time :/. It will probably also not happen in the future anymore, because I'm going to release Jedi 1.0 soon, so this is probably the last time for a long time that you have to deal with deprecations in Jedi.
> 
> Still wish you a Merry Christmas!

thanks a lot

**Comment by LuanComputacao** [comment]
pip install -U jedi==0.17.2 parso==0.7.1

**Comment by mostealth** [comment]
I wonder, when you have a dependency (jedi) that is in version 0.x.y (not reached major 1), if it is not wiser to pin the minor as >=0.17,<0.18.

**Comment by yurzo** [comment]
FYI https://github.com/ipython/ipython/pull/12751#issuecomment-758065462

**Comment by violet4** [comment]
> As a temporary fix for anyone just trying to get things working again:
> 
> ```
> pip install jedi==0.17.2
> ```
> 
> It would be really nice if you could quickly release a 7.19.1. (It's already fixed on master).
> 
> Sorry for that. I did not realize that IPython with that fix was not released yet. I usually test IPython completions before doing a Jedi release, but not this time :/. It will probably also not happen in the future anymore, because I'm going to release Jedi 1.0 soon, so this is probably the last time for a long time that you have to deal with deprecations in Jedi.
> 
> Still wish you a Merry Christmas!

note: in anaconda, this has to be done in the same kernel you're using from within the jupyter notebook, not the one you're using the `jupyter notebook` command from!

took me about 30 minutes to finally figure that one out once i found @davidhalter's pip install tip (i used `conda install jedi==0.17.2` instead of pip).

thank you all!

**Comment by gapster** [comment]
Can you explain what you mean by "same kernel ..." vs "using the jupyter notebook ..."   i.e.,  will conda install jedi==0.17.2 from the shell command line do the trick?  (it seems to)

**Comment by marvinbernhardt** [comment]
> Can you explain what you mean by "same kernel ..." vs "using the jupyter notebook ..." i.e., will conda install jedi==0.17.2 from the shell command line do the trick? (it seems to)

You can run Jupyter in one env and use it with a kernel from another env. See for example [this SO question](https://stackoverflow.com/questions/53004311/how-to-add-conda-environment-to-jupyter-lab). Jedi version of the latter env matters. If you are not aware of this, you probably use one conda env for jupyter and kernel.

**Comment by dowenk** [comment]
The following works in ipython/jupyter.   I assume it gets jedi out of the way of the built-in completer.   Good enough for me until fix is released.

%config Completer.use_jedi = False

**Comment by davidhalter** [comment]
Because people at this point probably don't scroll up to my comment (https://github.com/ipython/ipython/issues/12740#issuecomment-751273584), this is what you should do temporarily:

```
pip install jedi==0.17.2
```

**Comment by bl-ue** [comment]
When is the team going to fix this? In just a month hundreds of users have encountered this issue.

**Comment by bl-ue** [comment]
I guess pinning jedi to 0.172 should be added to the readme. Anyone think I should open a PR for that? 😸

**Comment by phinate** [comment]
just adding a +1 to this issue, it totally disables tab completion in Jupyter, so worth pinning 0.17.2 until resolved

**Comment by ghimireadarsh** [comment]
> As a temporary fix for anyone just trying to get things working again:
> 
> ```
> pip install jedi==0.17.2
> ```
> 
> It would be really nice if you could quickly release a 7.19.1. (It's already fixed on master).
> 
> Sorry for that. I did not realize that IPython with that fix was not released yet. I usually test IPython completions before doing a Jedi release, but not this time :/. It will probably also not happen in the future anymore, because I'm going to release Jedi 1.0 soon, so this is probably the last time for a long time that you have to deal with deprecations in Jedi.
> 
> Still wish you a Merry Christmas!

Thanks

**Comment by bl-ue** [comment]
This issue should be pinned on the GitHub repo.  @Carreau?

**Comment by rsokl** [comment]
I am curious about what the timeline looks like for this getting fixed for mainline conda users? 

I ask because I have an online outreach course spinning up in about four days, meaning that 600+ highschool students are going to be encountering the "now hit `<TAB>` to autocomplete 😃" portion of their virtual course, and it is going unceremoniously bomb them out of `ipython`.

Downgrading `jedi` is easy enough for me, but much harder for someone who has never written a line of code before.

**Comment by bnavigator** [comment]
Releasing a new Jedi version might be easier than waiting for the next IPython release.

@davidhalter, how about making a one time exception in this case and reintroducing the deprecated features, the lack of which breaks IPython (and a few more packages) in a Jedi 0.18.1?

**Comment by bnavigator** [comment]
> What the timeline looks like for this getting fixed for mainline conda users?

You might be sucessful if you ask the conda maintainers to yank jedi 0.18 from their repos.

**Comment by h-vetinari** [comment]
> You might be successful if you ask the conda maintainers to yank jedi 0.18 from their repos.

The last `ipython` builds on conda-forge [already](https://github.com/conda-forge/ipython-feedstock/blob/master/recipe/meta.yaml#L33-L34) depend on `jedi<0.18`, so this this shouldn't even happen if you create a new environment (or do `conda update ipython` first).

**Comment by rsokl** [comment]
>The last ipython builds on conda-forge already depend on jedi<0.18, so this this shouldn't even happen if you create a new environment (or do conda update ipython first).

Presumably if you have set conda-forge as the default channel, correct?

**Comment by h-vetinari** [comment]
Even Anaconda [followed](https://github.com/AnacondaRecipes/aggregate/commit/fdd96f6968d7d4b6b8ec68ff0caa4d8566e953f9#diff-b85a555c399c3f7c88b8c090a978f34b985a1b4d23db5d0fd8887dfc80b476d0) with this already, so that shouldn't matter.

**Comment by rsokl** [comment]
Hmm.. Then why is it that creating a new conda environment (with `conda 4.9.2`) giving me:

```
>>> conda create -n test_it python=3.8 ipython
The following NEW packages will be INSTALLED:
...
ipython            pkgs/main/win-64::ipython-7.19.0-py38hd4e2768_0
...
jedi               pkgs/main/win-64::jedi-0.18.0-py38haa95532_1
```

Am I missing something here?

**Comment by h-vetinari** [comment]
So the packages are definitely [there](https://anaconda.org/main/ipython/files) (notice the `_1` build number), but interestingly, are not seeing much (or even any) downloads. I don't know if your anaconda channel is behind a proxy somewhere (or indeed, if anaconda themselves do some more integration testing before a package goes "live"), but in any case, you can work around this by adding `conda config --add channels conda-forge`.

Side note: For any user that's not under the strictest of regulatory requirements (where people want to be able to blame a provider like Anaconda if stuff goes bad), conda-forge is more than ready/stable/secure to be your default channel. But, fair warning, that's just my subjective opinion.

**Comment by bnavigator** [comment]
> Am I missing something here?

Not you, but conda's dependency resolver. This is a problem and has been for years.  (Sorry for the off-topic, but I could not resist)

**Comment by bnavigator** [comment]
> So the packages are definitely [there](https://anaconda.org/main/ipython/files) (notice the `_1` build number), but interestingly, are not seeing much (or even any) downloads

My totally uneducated guess (because I never understood the internals of the conda resolver):
conda sees the conflict for `jedi==0.18.0 and ipython-7.19.0-..._1` and instead of resolving it by `jedi==0.17.2 and ipython-7.19.0-..._1` it chooses `jedi==0.18.0 and ipython-7.19.0-..._0`

**Comment by h-vetinari** [comment]
Welp, spoke too soon, conda-forge has the same problem (basically, the higher build number seems to lose against having one dependency have a higher version, which is not how it _should_ work. A possible solution would be backporting https://github.com/ipython/ipython/commit/dcd9dc90aee7e4c5c52ce44c18e7518934790612 to the feedstock, but that really should happen here first).

> Not you, but conda's dependency resolver. This is a problem and has been for years. (Sorry for the off-topic, but I could not resist)

Have you heard of [mamba](https://github.com/mamba-org/mamba) already? It has a rewritten resolver but is fully compatible with the conda-ecosystem. Not only does it solve the biggest problem of the conda-solver (speed), but it also correctly sets up the environment here:

```
>mamba create -n test python=3.8 ipython
[...]
  ipython                             7.19.0  py38hc5df569_2      conda-forge/win-64       1 MB
  ipython_genutils                     0.2.0  py_1                conda-forge/noarch      21 KB
  jedi                                0.17.2  py38haa244fe_1      conda-forge/win-64     944 KB
[...]
```

**Comment by rsokl** [comment]
Thank you, both of you, for the useful insights and recommendations. I really appreciate it.

It sounds like I will go the route of putting a call-out box in my website that adds a hand-holding jedi-downgrading walkthrough (although, then I have to hope that the students have the awareness to repeat that step when creating new environments)

As it stands, I am dangerously close to making *two* "but think of the children" pleas in a single github issue, which is a violation of one of my new years resolutions.

(mamba looks cool!)

**Comment by h-vetinari** [comment]
@rsokl
I raised an issue with conda itself, since it's probably too much (though not impossible) to do this on the conda-forge packaging side (in a way that's also compatible across several jedi-versions) - raised an [issue](https://github.com/conda-forge/ipython-feedstock/issues/127) to discuss that too...

@Carreau [said](https://github.com/ipython/ipython/pull/12751#issuecomment-758065462) he'd make a new release soon, which would of course also solve things.

In the meantime, the "hand-holding" guide should be as easy as `conda install ipython "jedi<0.18"` (the quotes are important though, as otherwise you end up piping to nowhere).

**Comment by h-vetinari** [comment]
@rsokl
There was one other way I hadn't thought about - fixing the conda-forge repodata directly. This was done yesterday, and means the problem is fixed _within_ conda-forge, i.e. when using strict channel priority (ignoring the main anaconda package that still has the same problem) - which unfortunately is not the out-of-the-box behaviour for conda though.

You can either do this for a single install (e.g. `conda create -n test --strict-channel-priority python=3.8 ipython`), but the recommended way (see e.g. [here](https://conda-forge.org/)) is persisting that setting with
```
conda config --set channel_priority strict
```

**Comment by gkuenning** [comment]
> Thank you, both of you, for the useful insights and recommendations. I really appreciate it.
> 
> It sounds like I will go the route of putting a call-out box in my website that adds a hand-holding jedi-downgrading walkthrough (although, then I have to hope that the students have the awareness to repeat that step when creating new environments)
> 
> As it stands, I am dangerously close to making _two_ "but think of the children" pleas in a single github issue, which is a violation of one of my new years resolutions.
> 
> (mamba looks cool!)

Can yuo post a link to your walkthrough?  I think some others among us have the same problem...

**Comment by gkuenning** [comment]
BTW, "Effective Python" (2nd edition) #89 suggests using the warnings module to handle things like deprecating arguments.

**Comment by rsokl** [comment]
@gkuenning I just uploaded the downgrade instructions to Python Like You Mean It.

Under: [Information Introduction to Python](https://www.pythonlikeyoumeanit.com/Module1_GettingStartedWithPython/Informal_Intro_Python.html) and under [Jupyter Notebooks](https://www.pythonlikeyoumeanit.com/Module1_GettingStartedWithPython/Jupyter_Notebooks.html)

**Comment by h-vetinari** [comment]
@rsokl 
Actually, anaconda fixed their repo data as well, so now it works out of the box again 🥳

**Comment by rsokl** [comment]
Nice! I just confirmed this on my end. I guess I'll keep those callout boxes up on PLYMI for a bit in case folks had already installed anaconda.

**Comment by Carreau** [comment]
I'll try to release 7.20 soon; I've make it compatible with jedi 0.18 as pip also use a resolver and so pinning is not an option as pip would be free to downgrade IPython and keep jedi 0.18 which will brake. 

Right now the limiting factor for the release is writing the what's new.

**Comment by mostealth** [comment]
> I'll try to release 7.20 soon; I've make it compatible with jedi 0.18 as pip also use a resolver and so pinning is not an option as pip would be free to downgrade IPython and keep jedi 0.18 which will brake.
> 
> Right now the limiting factor for the release is writing the what's new.

I am really disappointed that the maintainer of a great product as Ipython has such a limited knowledge about the problem.
The fact that Ipython has not been pinning jedi means that any version of it released in the past that depends on jedi is at its merci to stop working as soon as a no backwards compatible change is introduced and used in Ipython.
By changing Ipython to make it jedi 0.18 compatible and releasing 7.20 you make that 7.20 operational today. If tomorrow jedi makes another backwards incompatible change We will be screwed again.
And you can't blame jedi for backwards incompatible changes because it is not even 1.x and you are not even pinning the major.
You are calling for trouble.

**Comment by Carreau** [comment]
> I am really disappointed that the maintainer of a great product as Ipython has such a limited knowledge about the problem.
> The fact that Ipython has not been pinning jedi means that any version of it released in the past that depends on jedi is at its merci to stop working as soon as a no backwards compatible change is introduced and used in Ipython.
> By changing Ipython to make it jedi 0.18 compatible and releasing 7.20 you make that 7.20 operational today. If tomorrow jedi makes another backwards incompatible change We will be screwed again.
> And you can't blame jedi for backwards incompatible changes because it is not even 1.x and you are not even pinning the major.
> You are calling for trouble.

I'm amazed that someone that contribute squat is telling me what i'm doing wrong without tryign to really think about the problem.

- Say I pin IPython 7.20 to jedi <0.19 , and david release jedi 0.19. 
- pip is no free to say "hey let's install jedi 0.19 and downgrade IPython to 5.x". 
- Bam you are wrong™️ . 💎 🚀 🌕 

And second the change of API were in jedi 0.17 with deprecation warnings and fix in master close to a year ago; I thought that by now 8.0 would be out. 

FYI about me know nothing about versioning:
  - I spend month implementing on pushing the `python_requires` logic in pypi, and in pip to make sure when IPython dropped it did not break for Python 2 users
  - I'm one of the regular advocate to have version medata data and package content to be separated in conda/pypi for this exact reason.
  - It's not because something seem obvious that it's correct.

You could have just asked in for a question "why are not you pinning to <0.19 starting with IPython 7.20", to which I would have responded.
So not only you are impolite and haven't done your research,  everybody see you incorrectly assume you know better than others and  you've lost all credibility.

✋ ⬇️ 🎤

**Comment by yurzo** [comment]
@Carreau: dude, you rock!

**Comment by mostealth** [comment]
So essentially you say as there are problems with pip resolver and in some cases it will not be able to upgrade correctly I will completely ignore the problem and not try to solve other uses cases that I can.
Yes, fixing Ipython to work with the latest jedi is correct and pinning is needed as well.
Most use cases of Ipython will be install Ipython on a new environment. Should you pin, your 7.20 will work always. Ignore it and you depend on your dependencies.
The reason I didn't ask why you are planning to do what you are planning to do is because I know that solution is suboptimal.
It doesn't matter what work or knowledge has been accumulated in the past, that says nothing.

**Comment by dlukes** [comment]
> pip is no[w] free to say "hey let's install jedi 0.19 and downgrade IPython to 5.x".

That would indeed be unfortunate, the resulting setup would be not only broken, but out-of-date to boot. Can that happen when running `pip install ipython` in a fresh environment? I would naively assume that this would fetch the latest version of IPython and figure out the dependencies from there, i.e. honoring `jedi < 0.19`. But that assumption is probably wrong?

As maintainer of my own small Python package with moderately complicated dependencies, I feel like I should read up on this :) Any tips on good material on this topic?

**Comment by cknoll** [comment]
@mostealth Obviously you have a different perspective on that matter. That's fine. However, it seems rather unproductive to dissipate the time and energy of productive people like @Carreau with pointless discussions. If you have deeper insights I see the following options: You can fork the project, file a PR or just sate your improvement suggestion in a way that is arguable, maybe in a dedicated issue.

For the rest of us I would suggest to basically ignore distracting comments which contain neither helpful information nor constructive criticism, but instead spread insults and negative emotion. Reactinv with :-1: should be enough attention for such comments.

**Comment by davidhalter** [comment]
@Carreau Thanks a lot! I guess the next release is going to be 1.0 anyway, so it shouldn't happen again. We might then think about pinning `jedi < 2.0.0` (which is probably years away).

In general for all people like @mostealth: If you are worried about older versions breaking: **Pin your dependencies!** The Python packaging system is still in a bit of a problematic state. The devs are doing a good job trying to turn it around, but there is a lot of technical debt from years ago. Keep in mind they want to keep backwards compatibility with a lot of old and new things. Just a enumerating a few things to show how complicated it is: eggs, wheels, explicit namespace packages, setuptools, distutils, import hooks, import meta paths, importlib, easy_install, venvs, virtualenvs, zip imports, freezed packages, source packages for C/C++ libraries and a lot of other things that I don't even know or remember anymore.

Most of you guys complaining don't really understand how hard certain things are if you consider all consequences. Read about Chesterton's Fence and try to understand why you shouldn't want something changed unless you understand why it is that way in the first place.

**Comment by bnavigator** [comment]
Thank you @Carreau. Finally someone who does not follow the church of pinning stuff without thinking about the consequences. Here is another case why pinning to older versions is not always a good idea: https://github.com/palantir/python-language-server/issues/898#issuecomment-770399215.

Unfortunately the Jedi and Parso pair are a frequent cause for headaches for packagers. Not only pip and conda, but also Linux distributions. I am looking forward to a more stable API once they reach 1.0.

**Comment by Carreau** [comment]
Thanks everyone for the inputs; if we want to have discussion as to why/when pinning and why when not to pin, and other I suggest we don't use the bug tracker, maybe discourse.jupyter.org; but stop adding to this particular issue. 

IPython is a bit in a tough spot as it is both a library and application (in general you want to have a lock file to pin deps for applications, but be as liberal as possible when building a library). 

There is also a lot of issues that arise from the fact that maintainers time and desire to maintain can come and go; I definitely thought that by sept 2020 IPython 8 would be out, and that other maintainers like @davidhalter from jedi would also welcome more testing and help.

The Python package ecosystem is also quite old; and PyPI relatively recent; the requirements and needs of today's infrastructure can also be much different than what was expected even 5 years ago. That plus it works from HPC system to end-user  that are starting to program so there is a lot of difficulties there.

Thanks everyone, I've now released 7.20 feel free to open bug reports if you encounter any issues. 
For now I'm going to lock this thread to avoid getting much more off topic.
